Encoding/decoding negative integers

killerstorm opened the pull request:
I added support for encoding & decoding negative integers. Existing implementation was producing erroneous values.

I've added tests for numbers at the edges: -1, -128, -129. No extensive testing was done.

indutny commented on lib/asn1/decoders/der.js:
```diff
@@ -244,6 +244,10 @@ DERNode.prototype._decodeInt = function decodeInt(buffer, values) {
   var raw = buffer.raw();
   var res = new bignum(raw);
 
+  if ((raw[0] & 0x80) > 0) { // negative
```
Should it be `!== 0`? Please consider putting the comment on the line before `if`.

indutny replied:
Another note: should we rather use `bignum.fromTwos` here, and avoid creating `new bignum` if the result is negative?

killerstorm replied:
As far as I can tell, bignum.fromTwos doesn't exist.

indutny commented on lib/asn1/encoders/der.js:
```diff
@@ -195,10 +196,26 @@ DERNode.prototype._encodeInt = function encodeInt(num, values) {
+      // check if sign bit is occupied
+      if ((bitLength % 8) == 0) {
+	// when number is of form  10..0 no correction is needed
+	if (num.zeroBits() + 1 != num.bitLength()) {
```
Please use spaces for indent.

indutny replied:
`!==`

indutny commented on lib/asn1/encoders/der.js:
```diff
@@ -195,10 +196,26 @@ DERNode.prototype._encodeInt = function encodeInt(num, values) {
+      var bitLength = num.bitLength();
+      var width = Math.ceil(bitLength / 8) * 8;
+      // check if sign bit is occupied
+      if ((bitLength % 8) == 0) {
```
`===`

indutny commented on lib/asn1/encoders/der.js:
```diff
@@ -195,10 +196,26 @@ DERNode.prototype._encodeInt = function encodeInt(num, values) {
+	}
+      }
+      numArray = num.toTwos(width).toArray();
+      //console.log(numArray);
```
Please remove all commented code.

indutny commented:
I actually just added eslint to the `package.json`. So you can run `npm run lint` to fix the most of these issues.

killerstorm commented:
Thx, I also want to fix encoding JS integers > 2^32-1 while we are here. If you don't mind I'll redirect them to BN.

killerstorm commented:
Hi, I fixed lint issues and also added support for integers >= 2^31.

I don't see BN.fromTwos in BN library so I don't see an opportunity to handle it in a more efficient way.

Note that test "should encode OCSP request" fails, but it fails without my changes too.

blmalone commented:
Is this going to be merged?

Kalmac commented:
Is this only a conflict issue ? I need number support > 32 bits.
